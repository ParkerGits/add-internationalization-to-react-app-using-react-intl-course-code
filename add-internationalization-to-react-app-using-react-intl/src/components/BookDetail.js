import React from 'react';
import { 
  FormattedMessage,
  FormattedDate,
  FormattedTime,
  FormattedRelativeTime
} from 'react-intl' 
import {meanBy, round, sortBy} from 'lodash';

import books from '../books.json';

const BookDetail = ({match}) => {
  const book = books.find(book => book.id === parseInt(match.params.bookId, 10));
  const sortedReviews = sortBy(book.reviews, 'date').reverse();
  const avgRating = round(meanBy(book.reviews, (r) => r.rating), 2)

  return (
    <div className="BookDetail">
      <div className="BookDetail-meta">
        <img src={book.image} width="200" height="275" alt={book.title}/>
        <div className="BookDetail-metaBody">
          <h1>{book.title}</h1>
          <h3>
            <FormattedMessage id="detail.author" values={{author: book.author}} />
          </h3>
          <div>
            <input type="checkbox" id="toggle" hidden/>
            <p>{book.description}</p>
            <label className="BookDetail-descriptionToggle" htmlFor="toggle">
              <FormattedMessage id="detail.toggle" />
            </label>
          </div>
        </div>
      </div>

      <h3 className="BookDetail-merchantHeading">
          <FormattedMessage id="detail.purchase"/>
      </h3>
      <div className="BookDetail-merchants">
        {book.merchants.map((merchant) => (
          <a href={merchant.link} className="Merchant" key={merchant.name} target="_blank">
            <img src={merchant.icon} width="32" height="32" alt={merchant.name}/>
            <strong>{merchant.name}</strong>
            <p>{merchant.price}</p>
          </a>
        ))}
      </div>
      <FormattedMessage id="detail.window" values={{numMerchants: book.reviews.length, small: chunks => <small>{chunks}</small>, em: chunks => <em>{chunks}</em>}} />
      <h2>
          <FormattedMessage id="detail.reviewsHeading"/>
      </h2>
      <h3>
        <FormattedMessage id="detail.averageRating" values={{avg: avgRating}} /> ({book.reviews.length} Reviews)
      </h3>
      <div className="BookDetail-reviews">
        {sortedReviews.map((review) => (
          <div className="Review" key={review.date}>
            <div className="Review-meta">
              <img src={review.avatar} alt="Avatar"/>
              <p>
                <FormattedMessage id="detail.userRating" values={{name: <strong>{review.name}</strong>, rating: review.rating}}/>
                <br />
                {/* <FormattedDate
                  value={new Date(review.date)}
                  year="2-digit"
                  month="2-digit"
                  day="2-digit" /><br /> */}

                <FormattedTime
                  value={new Date(review.date)}
                  year="2-digit"
                  month="2-digit"
                  day="2-digit" /><br />
                <FormattedRelativeTime value={-1 * Math.floor((Date.now() - review.date)/((1000*60*60*24)))} numeric="auto" unit="days"/><br />
              </p>
            </div>
            <p>{review.body}</p>
          </div>
        ))}
      </div>
    </div>
  )
}

export default BookDetail;