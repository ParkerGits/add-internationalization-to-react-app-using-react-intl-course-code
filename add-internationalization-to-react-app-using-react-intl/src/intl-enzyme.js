/**
 * Components using the react-intl module require access to the intl context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * English-locale intl context around them.
 */

import React from 'react';
import {IntlProvider} from 'react-intl';
import {mount, shallow} from 'enzyme';
import {flattenMessages} from './utils'
// You can pass your messages to the IntlProvider. Optional: remove if unneeded.
import messages from './messages'
const defaultLocale = 'en-US';
const locale = defaultLocale;
const flattenedMessages = flattenMessages(messages[locale])


export function mountWithIntl(node) {
  return mount(node, {
    wrappingComponent: IntlProvider,
    wrappingComponentProps: {
      locale,
      defaultLocale,
      flattenedMessages,
    },
  });
}

export function shallowWithIntl(node) {
  return shallow(node, {
    wrappingComponent: IntlProvider,
    wrappingComponentProps: {
      locale,
      defaultLocale,
      flattenedMessages,
    },
  });
}