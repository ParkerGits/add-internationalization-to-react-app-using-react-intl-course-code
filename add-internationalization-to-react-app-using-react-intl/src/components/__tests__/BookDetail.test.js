import React from "react";
import { mountWithIntl, shallowWithIntl } from "../../intl-enzyme.js";
import toJson from "enzyme-to-json";

import BookDetail from "../BookDetail";

import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

configure({ adapter: new Adapter() })

Date.now = jest.fn(() => 1491111687199)

let wrapper = null;
describe('BookDetail', () => {
  it('renders expected markup', () => {
    wrapper = mountWithIntl(<BookDetail match={{params: {bookId: 1}}}/>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('intl messages', () => {
    beforeEach(() => {
      wrapper = (shallowWithIntl(<BookDetail match={{params: {bookId: 1}}}/>)).shallow().shallow()
    });

    it('renders a single "purchase" message', () => {
      expect(wrapper.find({id: 'detail.purchase'})).toHaveLength(1);
    });

    it('passes number of merchants to "new window" message', () => {
      expect(wrapper.find({id: 'detail.window'}).prop('values')).toMatchObject({
        numMerchants: 3
      });
    });
  });
});