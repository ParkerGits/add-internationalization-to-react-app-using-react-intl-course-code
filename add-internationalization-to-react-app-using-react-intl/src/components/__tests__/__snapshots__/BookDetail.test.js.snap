// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetail renders expected markup 1`] = `
<injectIntl(BookDetail)
  match={
    Object {
      "params": Object {
        "bookId": 1,
      },
    }
  }
>
  <BookDetail
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en-US",
        "formatDate": [Function],
        "formatDateToParts": [Function],
        "formatDisplayName": [Function],
        "formatList": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatNumberToParts": [Function],
        "formatPlural": [Function],
        "formatRelativeTime": [Function],
        "formatTime": [Function],
        "formatTimeToParts": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getDisplayNames": [Function],
          "getListFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralRules": [Function],
          "getRelativeTimeFormat": [Function],
        },
        "locale": "en-US",
        "messages": Object {},
        "onError": [Function],
        "textComponent": Symbol(react.fragment),
        "timeZone": undefined,
        "wrapRichTextChunksInFragment": undefined,
      }
    }
    match={
      Object {
        "params": Object {
          "bookId": 1,
        },
      }
    }
  >
    <div
      className="BookDetail"
    >
      <div
        className="BookDetail-meta"
      >
        <img
          alt="Dark Matter"
          height="275"
          src="https://images.gr-assets.com/books/1472119680l/27833670.jpg"
          width="200"
        />
        <div
          className="BookDetail-metaBody"
        >
          <h1>
            Dark Matter
          </h1>
          <h3>
            <FormattedMessage
              id="detail.author"
              values={
                Object {
                  "author": "Blake Crouch",
                }
              }
            >
              detail.author
            </FormattedMessage>
          </h3>
          <div>
            <input
              hidden={true}
              id="toggle"
              type="checkbox"
            />
            <p>
              “Are you happy with your life?”

Those are the last words Jason Dessen hears before the masked abductor knocks him unconscious.

Before he awakens to find himself strapped to a gurney, surrounded by strangers in hazmat suits.

Before a man Jason’s never met smiles down at him and says, “Welcome back, my friend.”

In this world he’s woken up to, Jason’s life is not the one he knows. His wife is not his wife. His son was never born. And Jason is not an ordinary college physics professor, but a celebrated genius who has achieved something remarkable. Something impossible.

Is it this world or the other that’s the dream? And even if the home he remembers is real, how can Jason possibly make it back to the family he loves? The answers lie in a journey more wondrous and horrifying than anything he could’ve imagined—one that will force him to confront the darkest parts of himself even as he battles a terrifying, seemingly unbeatable foe.

From the author of the bestselling Wayward Pines trilogy, Dark Matter is a brilliantly plotted tale that is at once sweeping and intimate, mind-bendingly strange and profoundly human—a relentlessly surprising science-fiction thriller about choices, paths not taken, and how far we’ll go to claim the lives we dream of.
            </p>
            <label
              className="BookDetail-descriptionToggle"
              htmlFor="toggle"
            >
              <FormattedMessage
                id="detail.toggle"
              >
                detail.toggle
              </FormattedMessage>
            </label>
          </div>
        </div>
      </div>
      <h3
        className="BookDetail-merchantHeading"
      >
        <FormattedMessage
          id="detail.purchase"
        >
          detail.purchase
        </FormattedMessage>
      </h3>
      <div
        className="BookDetail-merchants"
      >
        <a
          className="Merchant"
          href="https://www.amazon.com/Dark-Matter-Novel-Blake-Crouch/dp/1101904224"
          key="Amazon"
          target="_blank"
        >
          <img
            alt="Amazon"
            height="32"
            src="https://s-media-cache-ak0.pinimg.com/564x/cc/77/ef/cc77efac50fd7637763ba7115bc4f17a.jpg"
            width="32"
          />
          <strong>
            Amazon
          </strong>
          <p>
            <FormattedNumber
              currency="USD"
              currencyDisplay="symbol"
              style="currency"
              value="16.19"
            >
              $16.19
            </FormattedNumber>
          </p>
        </a>
        <a
          className="Merchant"
          href="http://www.barnesandnoble.com/w/dark-matter-blake-crouch/1122954090"
          key="Barnes & Noble"
          target="_blank"
        >
          <img
            alt="Barnes & Noble"
            height="32"
            src="http://www.barnesandnobleinc.com/wp-content/themes/bninc/assets/images/favicon.ico"
            width="32"
          />
          <strong>
            Barnes & Noble
          </strong>
          <p>
            <FormattedNumber
              currency="USD"
              currencyDisplay="symbol"
              style="currency"
              value="16.19"
            >
              $16.19
            </FormattedNumber>
          </p>
        </a>
        <a
          className="Merchant"
          href="http://www.audible.com/pd/Sci-Fi-Fantasy/Dark-Matter-Audiobook/B01CUKULGA"
          key="Audible"
          target="_blank"
        >
          <img
            alt="Audible"
            height="32"
            src="http://logo.clearbit.com/audible.com"
            width="32"
          />
          <strong>
            Audible
          </strong>
          <p>
            <FormattedNumber
              currency="USD"
              currencyDisplay="symbol"
              style="currency"
              value="16.19"
            >
              $16.19
            </FormattedNumber>
          </p>
        </a>
      </div>
      <FormattedMessage
        id="detail.window"
        values={
          Object {
            "em": [Function],
            "numMerchants": 3,
            "small": [Function],
          }
        }
      >
        detail.window
      </FormattedMessage>
      <h2>
        <FormattedMessage
          id="detail.reviewsHeading"
        >
          detail.reviewsHeading
        </FormattedMessage>
      </h2>
      <h3>
        <FormattedMessage
          id="detail.averageRating"
          values={
            Object {
              "avg": 3.25,
              "count": 4,
            }
          }
        >
          detail.averageRating
        </FormattedMessage>
      </h3>
      <div
        className="BookDetail-reviews"
      >
        <div
          className="Review"
          key="1492668996053"
        >
          <div
            className="Review-meta"
          >
            <img
              alt="Avatar"
              src="http://i.pravatar.cc/64?img=2"
            />
            <p>
              <FormattedMessage
                id="detail.userRating"
                values={
                  Object {
                    "name": <strong>
                      Kirby Evans
                    </strong>,
                    "rating": 1,
                  }
                }
              >
                detail.userRating
              </FormattedMessage>
              <br />
              <FormattedTime
                day="2-digit"
                month="2-digit"
                value={2017-04-20T06:16:36.053Z}
                year="2-digit"
              >
                04/19/17, 11:16 PM
              </FormattedTime>
              <br />
              <FormattedRelativeTime
                numeric="always"
                style="short"
                unit="day"
                value={19}
              >
                in 19 days
              </FormattedRelativeTime>
              <br />
            </p>
          </div>
          <p>
            Ut irure excepteur in culpa. Elit amet enim qui laboris do labore duis tempor exercitation quis voluptate officia enim proident. Et ex eu magna elit et culpa quis duis esse fugiat. Aliqua esse reprehenderit non adipisicing deserunt nostrud commodo tempor. Incididunt labore id anim aliqua sit qui. Ex ad est officia ea esse et. Consectetur tempor nostrud et ea culpa.
          </p>
        </div>
        <div
          className="Review"
          key="1491743020373"
        >
          <div
            className="Review-meta"
          >
            <img
              alt="Avatar"
              src="http://i.pravatar.cc/64?img=3"
            />
            <p>
              <FormattedMessage
                id="detail.userRating"
                values={
                  Object {
                    "name": <strong>
                      Alicia Warner
                    </strong>,
                    "rating": 4,
                  }
                }
              >
                detail.userRating
              </FormattedMessage>
              <br />
              <FormattedTime
                day="2-digit"
                month="2-digit"
                value={2017-04-09T13:03:40.373Z}
                year="2-digit"
              >
                04/09/17, 6:03 AM
              </FormattedTime>
              <br />
              <FormattedRelativeTime
                numeric="always"
                style="short"
                unit="day"
                value={8}
              >
                in 8 days
              </FormattedRelativeTime>
              <br />
            </p>
          </div>
          <p>
            Dolor fugiat mollit do occaecat mollit aute incididunt esse cillum dolore commodo qui do minim. Aliqua proident aliqua ullamco sunt fugiat duis et exercitation deserunt non exercitation cupidatat id minim. Exercitation voluptate elit consectetur consectetur qui exercitation reprehenderit ut. Commodo laborum eiusmod sit nostrud sunt ea dolor est consequat enim est occaecat officia qui. Dolore nisi nulla laborum consectetur.
          </p>
        </div>
        <div
          className="Review"
          key="1491111687199"
        >
          <div
            className="Review-meta"
          >
            <img
              alt="Avatar"
              src="http://i.pravatar.cc/64?img=1"
            />
            <p>
              <FormattedMessage
                id="detail.userRating"
                values={
                  Object {
                    "name": <strong>
                      Gretchen Cannon
                    </strong>,
                    "rating": 5,
                  }
                }
              >
                detail.userRating
              </FormattedMessage>
              <br />
              <FormattedTime
                day="2-digit"
                month="2-digit"
                value={2017-04-02T05:41:27.199Z}
                year="2-digit"
              >
                04/01/17, 10:41 PM
              </FormattedTime>
              <br />
              <FormattedRelativeTime
                numeric="always"
                style="short"
                unit="day"
                value={-0}
              >
                in 0 days
              </FormattedRelativeTime>
              <br />
            </p>
          </div>
          <p>
            Occaecat ut ut mollit occaecat id fugiat velit est culpa. Nostrud incididunt commodo adipisicing duis cupidatat officia nisi cupidatat sit deserunt consectetur ex veniam. Pariatur qui in sit ut.
          </p>
        </div>
        <div
          className="Review"
          key="1489702832337"
        >
          <div
            className="Review-meta"
          >
            <img
              alt="Avatar"
              src="http://i.pravatar.cc/64?img=4"
            />
            <p>
              <FormattedMessage
                id="detail.userRating"
                values={
                  Object {
                    "name": <strong>
                      Pope Robbins
                    </strong>,
                    "rating": 3,
                  }
                }
              >
                detail.userRating
              </FormattedMessage>
              <br />
              <FormattedTime
                day="2-digit"
                month="2-digit"
                value={2017-03-16T22:20:32.337Z}
                year="2-digit"
              >
                03/16/17, 3:20 PM
              </FormattedTime>
              <br />
              <FormattedRelativeTime
                numeric="always"
                style="short"
                unit="day"
                value={-16}
              >
                16 days ago
              </FormattedRelativeTime>
              <br />
            </p>
          </div>
          <p>
            Esse consectetur exercitation sit qui minim commodo veniam tempor cupidatat qui fugiat occaecat sit. Magna fugiat non mollit commodo sit enim culpa. Pariatur reprehenderit est enim incididunt pariatur ullamco anim cupidatat labore. Duis magna id fugiat nulla exercitation ipsum.
          </p>
        </div>
      </div>
      <textarea
        placeholder="detail.inputPlaceholder"
      />
    </div>
  </BookDetail>
</injectIntl(BookDetail)>
`;
